README.md
# SiteAlert

This plugin monitors the health and security of your WordPress site! It check your WordPress installation to ensure that it is healthy, up to date, and secure. Use the SiteAlert page to quickly see the results of these checks using the simple color-coded sections.

Feel free to browse the code and make suggestions/requests.

Check out our [premium features](https://sitealert.io/) to get instant alerts if you site goes down, speed test monitoring, and much more!

## Getting Started

We do not currently use any build scripts. This repo can be downloaded as a zip and installed as-is as a WordPress plugin.

Once installed, there will be a new **SiteAlert** page added to the Tools menu and heart icon to your admin bar if the results of these checks need attention.

## Contributing

Community made feature requests, patches, localisations, bug reports and contributions are always welcome and are crucial to ensure SiteAlert continues to grow to become the #1 WordPress monitor.

When contributing please ensure you follow the guidelines below so that we can keep on top of things.

Please Note: GitHub is not intended for support based questions. For those, please use the [support forums](https://wordpress.org/support/plugin/my-wp-health-check).

### Creating Issues

* If you have any bugs or feature requests, please [create an issue](https://github.com/fpcorso/wordpress-health-check/issues/new)
* For bug reports, please clearly describe the bug/issue and include steps on how to reproduce it
* For feature requests, please clearly describe what you would like, how it would be used, and example screenshots (if possible)

### Pull Requests

* Ensure you stick to the [WordPress Coding Standards](https://codex.wordpress.org/WordPress_Coding_Standards)
* When committing, reference your issue (if present) and include a note about the fix
* Push the changes to your fork and submit a pull request to the 'master' branch of this repository
* We are trying to ensure that every function is documented well and follows the standards set by phpDoc going forward
* Finally, please use tabs and not spaces. The tab indent size should be 4 for all SiteAlert code

## Deployment

This WordPress plugin is automatically deployed when a new GitHub release is created from master.

## Versioning

We use [SemVer](http://semver.org/) for versioning. For the versions available, see [the releases in this repository](https://github.com/fpcorso/wordpress-health-check/releases).

## Developers

* SiteAlert team

See also [the list of contributors](https://github.com/fpcorso/wordpress-health-check/graphs/contributors) who participated in this project.

## License

This project is licensed under the GPLv2 License - see the [LICENSE.md file](https://github.com/fpcorso/wordpress-health-check/blob/master/LICENSE.md) for details.

## Support

This is a developer's portal for SiteAlert and should _not_ be used for support. Please create a support ticket [here](https://wordpress.org/support/plugin/my-wp-health-check).


CHANGELOG.MD
# SiteAlert WordPress Plugin Changelog

## 1.9.7 (Feb 20, 2022)

* Tested on WordPress 5.9

## 1.9.6 (Jan 8, 2022)

* Design improvements to checks and settings tabs
* Tested on WordPress 5.8.3

## 1.9.5 (October 26, 2021)

* Minor design tweaks

## 1.9.4 (October 20, 2021)

* Bump tested to up to WordPress 5.8

## 1.9.3 (March 14, 2021)

* Bump tested to up to WordPress 5.7
* Minor design tweaks

## 1.9.2 (December 21, 2020)

* Minor text changes
* Adds second screenshot

## 1.9.1 (December 7, 2020)

* Rewrite database version check to be easier to add new versions as they are released
* Add the database app name (MySQL or MariaDB) to REST API endpoint
* Add PHP 8.0 to PHP check
* Add new continuous deployment for plugin readme
* Increases spam comment threshold to 150 for check to fail

## 1.9.0 (Nov 23, 2020)

* Adds a new endpoint for upcoming email deliverability monitor
* Bump minimum WordPress version to 5.3, which is 12 months old
* Bump tested to up to WordPress 5.6
* Bump minimum PHP version to 5.6 to match WordPress

## 1.8.17 (Oct 16, 2020)

* Changes name to SiteAlert

## 1.8.16 (Oct 4, 2020)

* Makes minor improvements to settings tab JavaScript
* Changes API endpoint for when adding sites to premium account

## 1.8.15 (Sept 22, 2020)

* Bump minimum WordPress version to 5.1, which is 18 months old
* Fix bug in telemetry opt-in causing multiple requests on sites that opted-in

## 1.8.14 (Sept 6, 2020)

* Fix error notice relating to permission callback in REST API
* Move all api key authentication in REST API to new permission callback
* Added new upsell message

## 1.8.13 (August 31, 2020)

* Modify telemtry opt-in
* Add info about email summary premium feature

## 1.8.12 (August 5, 2020)

* Ensures WP Health works with WordPress 5.5

## 1.8.11 (April 27, 2020)

* Ensures WP Health works with the latest WordPress version

## 1.8.10 (February 28 , 2020)

* Temporarily disables vulnerability checker to due changes with 3rd party API
* Bumps tested WordPress version to recent version

## 1.8.9 (Sept 10, 2019)

* Bumps minimum WordPress version to 4.9 which is 2 years old
* Makes minor styling changes
* Adjusts review message

## 1.8.8 (July 25, 2019)

* Adds link for more information about blacklists

## 1.8.7 (July 16, 2019)

* Temporarily turns off review message

## 1.8.6 (June 21, 2019)

* Sorts the checks with failed checks at top of each section

## 1.8.5 (June 1, 2019)

* Changes text info for the API Key setting
* Makes minor design changes

## 1.8.4 (May 25, 2019)

* Increases spam comment threshold to 50 for check to fail
* Fixes minor capitalization and grammar mistakes
* Reduces refresh time for admin bar count from 1 hour to 45 minutes

## 1.8.3 (May 16, 2019)

* Closed Bug: Failed admin username check shows green ([Issue #112](https://github.com/fpcorso/wordpress-health-check/issues/112))

## 1.8.2 (May 8, 2019)

* Ensures compatibility with WordPress 5.2

## 1.8.1 (April 16, 2019)

* Increases number of spam messages needed to trigger check to fail

## 1.8.0 (April 1, 2019)

* Closed Enhancement: Check if site is discouraging search engines ([Issue #105](https://github.com/fpcorso/wordpress-health-check/issues/105))

## 1.7.9 (March 22, 2019)

* Makes minor changes to text

## 1.7.8 (March 14, 2019)

* Fixes bug causing ads to still show for premium users

## 1.7.7 (March 9, 2019)

* Makes minor changes to text

## 1.7.6 (February 22, 2019)

* Ensures compatibility with WordPress 5.1

## 1.7.5 (February 6, 2019)

* Improves internationalization

## 1.7.4 (January 22, 2019)

* Adds info on new blacklist monitoring feature to readme

## 1.7.3 (December 28, 2018)

* Modifies text for PHP check since two versions are ending support this month.

## 1.7.2 (December 8, 2018)

* Ensures plugin is compatible with WordPress 5.0

## 1.7.1 (November 26, 2018)

* Adds link to new article on why to update your site to relevant checks
* Adds info regarding new broken image monitoring in premium version to readme

## 1.7.0 (October 30, 2018)

* Adds new spam comments check to advise users to delete spam comments
* Adds value key to JSON object in the REST API

## 1.6.11 (October 18, 2018)

* Fixes plugin update check in REST API

## 1.6.10 (October 14, 2018)

* Makes minor text changes

## 1.6.9 (September 13, 2018)

* Enhances WordPress version check to work on hosts modifying the get_core_updates() function

## 1.6.8 (September 7, 2018)

* Adds link to checks from plugins page
* Makes minor text changes
* Improves internationalization

## 1.6.7 (August 31, 2018)

* Adds info about new premium features

## 1.6.6 (August 23, 2018)

* Fixes undefined variable bug in check for MariaDB

## 1.6.5 (August 13, 2018)

* Makes minor text changes
* Changes permission level to administrator instead of editor to see the checks

## 1.6.4 (July 30, 2018)

* Makes minor text changes

## 1.6.3 (July 7, 2018)

* Makes minor design and text changes
* Improves internationalization

## 1.6.2 (July 2, 2018)

* Changes name of plugin to WP Health
* Fixes incorrect offest error
* Makes minor design and text changes

## 1.6.1 (June 25, 2018)

* Fixes style issue with new settings tab

## 1.6.0 (June 24, 2018)

* Closed Bug: Plugin page not showing all unsupported plugins ([Issue #59](https://github.com/fpcorso/wordpress-health-check/issues/59))
* Closed Enhancement: External endpoint to run check ([Issue #50](https://github.com/fpcorso/wordpress-health-check/issues/50))

## 1.5.1 (June 21, 2018)

* Makes minor performance changes for sites with many plugins

## 1.5.0 (June 5, 2018)

* Closed enhancement: Add alerts on plugins page ([Issue #26](https://github.com/fpcorso/wordpress-health-check/issues/26))
* Closed bug: Isn't compatible with MariaDB ([Issue #44](https://github.com/fpcorso/wordpress-health-check/issues/44))

## 1.4.4 (May 24, 2018)

* Fixes bug causes undefined function errors on the frontend when logged in

## 1.4.3 (October 29, 2017)


php/wp_hc_admin.php
<?php

// Exit if accessed directly
if ( ! defined( 'ABSPATH' ) ) exit;

/**
 * Class To Display Admin Page
 *
 * @since 0.1.0
 */
class WPHC_Admin {
  /**
   * Main construct
   *
   * @since 0.1.0
   */
  function __construct() {
    $this->load_dependencies();
    $this->load_hooks();
  }

  /**
   * Load File Dependencies
   *
   * @since 0.1.0
   */
  private function load_dependencies() {
    ///None yet
  }

  /**
   * Adds Plugin's Functions To WordPress Hooks
   */
  private function load_hooks() {
    add_action( 'admin_menu', array( $this, 'setup_admin_page' ) );
  }

  /**
   * Creates the menu for the plugin
   *
   * @since 0.1.0
   */
  public function setup_admin_page() {
    add_management_page( 'WordPress Health Check', __( 'Health Check', 'my-wp-health-check' ), 'moderate_comments', 'wp-health-check', array( $this, 'settings_page' ) );
  }

  /**
   * Function to display the admin page
   *
   * @since 0.1.0
   */
  public function settings_page() {
    wp_enqueue_style( 'wp-hc-style', plugins_url( "../css/main.css", __FILE__ ) );
    ?>
    <div class="wrap">
  	   <h2>WordPress Health Check</h2>
       <p>If you feel that your website has benefited from this plugin, please help other users find this plugin by <a href="https://wordpress.org/support/view/plugin-reviews/my-wp-health-check">leaving a review</a>.</p>
       <hr />
       <h3>Server Check</h3>
       <?php
       $this->php_check();
       $this->mysql_check();
       do_action( 'wphc_server_check' );
       ?>
       <h3>WordPress Check</h3>
       <?php
       $this->wordpress_version_check();
       $this->admin_user_check();
       $this->themes_check();
       do_action( 'wphc_wordpress_check' );
       ?>
       <h3>Plugin Check</h3>
       <?php
       $this->update_plugins_check();
       $this->inactive_plugins_check();
       $this->supported_plugin_check();
       do_action( 'wphc_plugin_check' );
       ?>
    </div>
    <?php
  }

  /**
   * Checks if using latest version of WordPress
   *
   * @since 0.1.0
   */
  public function wordpress_version_check() {
    $core_update = get_core_updates();
    if ( ! isset( $updates[0]->response ) || 'latest' == $updates[0]->response ) {
      echo "<div class='wp-hc-good-box'><span class='dashicons dashicons-flag'></span>Your WordPress is up to date. Great Job!</div>";
    } else {
      echo "<div class='wp-hc-bad-box'><span class='dashicons dashicons-dismiss'></span>Your WordPress is not up to date. Your site has not received the latest security fixes and is less secure from hackers. Please consider updating.</div>";
    }
  }

  /**
   * Checks if using latest version of plugins
   *
   * @since 0.2.0
   */
  public function update_plugins_check() {
    $plugin_updates = get_plugin_updates();
    if( ! empty( $plugin_updates ) ) {
      $plugins = array();
      foreach ( $plugin_updates as $plugin ) {
        $plugins[] = $plugin->Name;
      }
      $plugin_list = implode( ",", $plugins );
      echo "<div class='wp-hc-bad-box'><span class='dashicons dashicons-dismiss'></span>You are not using the latest version of these plugins: $plugin_list. These updates could contain important security updates. Please update your plugins to ensure your site is secure and safe.</div>";
    } else {
      echo "<div class='wp-hc-good-box'><span class='dashicons dashicons-flag'></span>All of your WordPress plugins are up to date. Great Job!</div>";
    }
  }

  /**
   * Checks for inactive plugins
   *
   * @since 1.1.0
   */
  public function inactive_plugins_check() {

    // Gets all the plugins
    $plugins = get_plugins();
    $inactive_plugins = array();

    // Looks for any inactive plugins
    if ( ! empty( $plugins ) ) {
      foreach ( $plugins as $key => $plugin ) {
        if ( is_plugin_inactive( $key ) ) {
          $inactive_plugins[] = $plugin['Name'];
        }
      }
    }

    // If any plugins are inactive, display error message. If not, display success message
    if ( ! empty( $inactive_plugins ) ) {
      echo "<div class='wp-hc-bad-box'><span class='dashicons dashicons-dismiss'></span>These plugins are not active: " . implode( ', ', $inactive_plugins ) . ". Inactive plugins can still be compromised by hackers. If you are not using them, please uninstall them.</div>";
    } else {
      echo "<div class='wp-hc-good-box'><span class='dashicons dashicons-flag'></span>All of your plugins installed on the site are in use. Great job!</div>";
    }

  }

  /**
   * Checks For Unsupported Plugins
   *
   * Checks the installed plugins to see there is a plugin that hasn't been updated in over 2 years
   *
   * @since 1.0.0
   */
  public function supported_plugin_check() {
    $slugs = array();
    $unsupported_plugins = array();
    $plugin_info = get_site_transient( 'update_plugins' );
    if ( isset( $plugin_info->no_update ) ) {
      foreach ( $plugin_info->no_update as $plugin ) {
        $slugs[] = $plugin->slug;
      }
    }

    if ( isset( $plugin_info->response ) ) {
      foreach ( $plugin_info->response as $plugin ) {
        $slugs[] = $plugin->slug;
      }
    }
    foreach ($slugs as $plugin) {
      $response = wp_remote_get( "http://api.wordpress.org/plugins/info/1.0/$plugin" );
      $plugin_info = unserialize( $response['body'] );
      if ( time() - ( 60 * 60 * 24 * 365 * 2 ) > strtotime( $plugin_info->last_updated ) ) {
        $unsupported_plugins[] = $plugin_info->name;
      }
    }

    $plugin_list = implode(",", $unsupported_plugins);
    if ( empty( $unsupported_plugins ) ) {
      echo "<div class='wp-hc-good-box'><span class='dashicons dashicons-flag'></span>All of your plugins are currently supported. Great Job!</div>";
    } else {
      echo "<div class='wp-hc-bad-box'><span class='dashicons dashicons-dismiss'></span>The following plugins have not been updated in over two years which indicate that they are no longer supported by their developer: $plugin_list. There could be security issues that will not be fixed! Please reach out to the developers to ensure these plugins are still supported or look for alternatives and uninstall these plugins.</div>";
    }
  }

  /**
   * Checks if there is a user called admin
   *
   * @since 1.0.0
   */
  public function admin_user_check() {
    $user = get_user_by( 'login', 'admin' );
    if ( false === $user ) {
      echo "<div class='wp-hc-good-box'><span class='dashicons dashicons-flag'></span>Your site does not have a user 'admin'. Great job!</div>";
    } else {
      echo "<div class='wp-hc-bad-box'><span class='dashicons dashicons-dismiss'></span>There is a user 'admin' on your site. Hackers use this username when trying to gain access to your site. Please change this username to something else.</div>";
    }
  }

  /**
   * Checks if using latest version of themes
   *
   * @since 0.2.0
   */
  public function themes_check() {
    $theme_updates = get_theme_updates();
    if( ! empty( $theme_updates ) ) {
      $themes = array();
      echo "<div class='wp-hc-bad-box'><span class='dashicons dashicons-dismiss'></span>One or more of your themes have updates available. These updates could contain important security updates. Please update your plugins to ensure your site is secure and safe.</div>";
    } else {
      echo "<div class='wp-hc-good-box'><span class='dashicons dashicons-flag'></span>All of your WordPress themes are up to date. Great Job!</div>";
    }
  }

  /**
   * Checks if using latest version of mysql
   *
   * @since 0.1.0
   */
  public function mysql_check() {
    //Check for MySQL
    global $wpdb;
    $version = explode( '.', $wpdb->db_version() );
    $sql_check_health = 'good';
    $message = '';
    switch ( intval( $version[0] ) ) {
      case 4:
        $sql_check_health = 'bad';
        $message = "You server is running MySQL version " . $wpdb->db_version() . " which has not been supported in over 5 years and is below the required 5.0. Using an unsupported version of MySQL means that you are using a version that no longer receives important security updates and fixes. You must update your MySQL or contact your host immediately!";
        break;

      case 5:
        switch ( intval( $version[1] ) ) {
          case 0:
            $sql_check_health = 'okay';
            $message = "You server is running MySQL version " . $wpdb->db_version() . ". This is the bare minimum that WordPress requires. However, this version has not been supported in 2 years and is below the recommended 5.5. Using an unsupported version of MySQL means that you are using a version that no longer receives important security updates and fixes. You should consider updating your MySQL or contacting your host.";
            break;

          case 1:
            $sql_check_health = 'okay';
            $message = "You server is running MySQL version " . $wpdb->db_version() . ". This is above the bare minimum that WordPress requires. However, this version is not longer supported and below the recommended 5.5. Using an unsupported version of MySQL means that you are using a version that no longer receives important security updates and fixes. You should consider updating your MySQL or contacting your host.";
            break;

          case 5:
            $sql_check_health = 'good';
            $message = "You server is running MySQL version " . $wpdb->db_version() . ". Good job! This is the recommended version.";
            break;

          case 6:
            $sql_check_health = 'good';
            $message = "You server is running MySQL version " . $wpdb->db_version() . ". Good job! This is above the recommended version.";
            break;

          case 7:
            $sql_check_health = 'good';
            $message = "You server is running MySQL version " . $wpdb->db_version() . ". Good job! This is the latest version.";
            break;


          default:
            $sql_check_health = 'bad';
            $message = "Error checking MySQL health.";
            break;
        }
        break;

      default:
        $sql_check_health = 'bad';
        $message = "Error checking MySQL health.";
        break;
    }
    switch ( $sql_check_health ) {
      case 'good':
        echo "<div class='wp-hc-good-box'><span class='dashicons dashicons-flag'></span>$message</div>";
        break;

      case 'okay':
        echo "<div class='wp-hc-okay-box'><span class='dashicons dashicons-lightbulb'></span>$message</div>";
        break;

      case 'bad':
        echo "<div class='wp-hc-bad-box'><span class='dashicons dashicons-dismiss'></span>$message</div>";
        break;

      default:
        echo "<div class='wp-hc-bad-box'><span class='dashicons dashicons-dismiss'></span>Error checking PHP health.</div>";
        break;
    }
  }

  /**
   * Checks if using latest version of php
   *
   * @since 0.1.0
   */
  public function php_check() {
    $version = explode( '.', PHP_VERSION );
    $php_check_health = 'good';
    $message = '';
    switch ( intval( $version[0] ) ) {
      case 4:
        $php_check_health = 'bad';
        $message = "You server is running PHP version " . PHP_VERSION . " which has not been supported in over 5 years and is below the required 5.2. Using an unsupported version of PHP means that you are using a version that no longer receives important security updates and fixes. You must update your PHP or contact your host immediately!";
        break;

      case 5:
        switch ( intval( $version[1] ) ) {
          case 0:
            $php_check_health = 'bad';
            $message = "You server is running PHP version " . PHP_VERSION . " which has not been supported in almost 10 years and is below the required 5.2. Using an unsupported version of PHP means that you are using a version that no longer receives important security updates and fixes. You must update your PHP or contact your host immediately!";
            break;

          case 1:
            $php_check_health = 'bad';
            $message = "You server is running PHP version " . PHP_VERSION . " which has not been supported in almost 10 years and is below the required 5.2. Using an unsupported version of PHP means that you are using a version that no longer receives important security updates and fixes. You must update your PHP or contact your host immediately!";
            break;

          case 2:
            $php_check_health = 'okay';
            $message = "You server is running PHP version " . PHP_VERSION . ". This is the bare minimum requirement of WordPress. However, this version has not been supported in almost 5 years and is below the recommended 5.5. Using an unsupported version of PHP means that you are using a version that no longer receives important security updates and fixes. You should consider updating your PHP or contact your host.";
            break;

          case 3:
            $php_check_health = 'okay';
            $message = "You server is running PHP version " . PHP_VERSION . ". This is the above the bare minimum requirement of WordPress. However, this version has not been supported in almost 12 months and is below the recommended 5.5. Using an unsupported version of PHP means that you are using a version that no longer receives important security updates and fixes. You should consider updating your PHP or contact your host.";
            break;

          case 4:
            $php_check_health = 'okay';
            $message = "You server is running PHP version " . PHP_VERSION . ". This is the above the bare minimum requirement of WordPress. However, this version will no longer be supported as of September 2015. Check with your host to ensure they plan on updating before this version is no longer supported.";
            break;

          case 5:
            $php_check_health = 'good';
            $message = "You server is running PHP version " . PHP_VERSION . ". Good job! This is the recommended version.";
            break;

          case 6:
            $php_check_health = 'good';
            $message = "You server is running PHP version " . PHP_VERSION . ". Good job! This is above the recommended version.";
            break;

          default:
            $php_check_health = 'bad';
            $message = "Error checking PHP health.";
            break;
        }
        break;

      case 7:
        $php_check_health = 'good';
        $message = "You server is running PHP version " . PHP_VERSION . ". Good job! This is the latest version.";
        break;

      default:
        $php_check_health = 'bad';
        $message = "Error checking PHP health.";
        break;
    }
    switch ( $php_check_health ) {
      case 'good':
        echo "<div class='wp-hc-good-box'><span class='dashicons dashicons-flag'></span>$message</div>";
        break;

      case 'okay':
        echo "<div class='wp-hc-okay-box'><span class='dashicons dashicons-lightbulb'></span>$message</div>";
        break;

      case 'bad':
        echo "<div class='wp-hc-bad-box'><span class='dashicons dashicons-dismiss'></span>$message</div>";
        break;

      default:
        echo "<div class='wp-hc-bad-box'><span class='dashicons dashicons-dismiss'></span>Error checking PHP health.</div>";
        break;
    }
  }
}

$wp_hc_admin = new WPHC_Admin();
?>
